borg/platform.py

```diff
@@ -14,3 +18,86 @@ def acl_get(path, item, st, numeric_owner=False):
 
     def acl_set(path, item, numeric_owner=False):
         pass
+
+# GenericDirEntry, scandir_generic (c) 2012 Ben Hoyt
+# from the python-scandir package (3-clause BSD license, just like us, so no troubles here)
+# note: simplified version
+
+
+class GenericDirEntry(object):
+    __slots__ = ('name', '_stat', '_lstat', '_scandir_path', '_path')
+
+    def __init__(self, scandir_path, name):
+        self._scandir_path = scandir_path
+        self.name = name
+        self._stat = None
+        self._lstat = None
+        self._path = None
+
+    @property
+    def path(self):
+        if self._path is None:
+            self._path = os.path.join(self._scandir_path, self.name)
+        return self._path
+
+    def stat(self, follow_symlinks=True):
+        assert not follow_symlinks
+        return os.lstat(self.path)
+
+    def is_dir(self, follow_symlinks=True):
+        assert not follow_symlinks
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink#
+        return stat.S_ISDIR(st.st_mode)
+
+    def is_file(self, follow_symlinks=True):
+        assert not follow_symlinks
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink
+        return stat.S_ISREG(st.st_mode)
+
+    def is_symlink(self):
+        try:
+            st = self.stat(False)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+            return False  # Path doesn't exist or is a broken symlink
+        return stat.S_ISLNK(st.st_mode)
+
+    def inode(self):
+        st = self.stat(False)
+        return st.st_ino
```

Comment:
> This is a shim implementation of the scandir interfaces; can't change the API here. Since we never want to follow symlinks I added the asserts and hardcoded lstat.

That's the code used if this is running on neither Python 3.5 (which includes scandir) or Python 3.4 with python-scandir installed.